Feature: Whitelist overhaul

SakuraIsayeki opened the pull request:
Overhauled the whitelist system to use a new dedicated service. 
This one is memory-based to optimise lookups, rather than accessing the file at each server join for WL comparison. 

Added support for in-file comments, by using `#` prefix (considered invalid and ignored by parser). Could come in handy for those massive files, or just keeping track of "whose IP is this" cases.
Added support for ranges using CIDR notation, to allow for entire subnets to be whitelisted. Particularly useful for LAN parties. 

New default for the `whitelist.txt` file looks like this : 
```ini
# Localhost
127.0.0.1
::1

# Uncomment to allow IPs within private ranges
# 10.0.0.0/8
# 172.16.0.0/12
# 192.168.0.0/16
# fe80::/10
# fd00::/8
```

Resolves #3091.

SakuraIsayeki committed:
feat: implement IP whitelist management service

Introduces a class for managing a whitelist of IP addresses and networks, enabling adding, removing, and persisting entries to a file. Whitelist is used to allow or restrict access based on IP, with support for CIDR networks and individual IPs.

SakuraIsayeki committed:
style: update IP comments for private ranges

Clarifies and adds more private IP range options in the whitelist comments for better configuration guidance

SakuraIsayeki committed:
feat: Use new whitelist system

Introduces the whitelist system as a core component and applies it during player IP checks, replacing direct file-based whitelisting calls with a dedicated service to improve clarity and maintainability.

SakuraIsayeki committed:
refactor: Improve default whitelist file creation

SakuraIsayeki committed:
refactor: move whitelist management to static property

SakuraIsayeki committed:
feat: use new whitelist service for add command

SakuraIsayeki committed:
fix: ensure newlines before appending to file

Adds logic to check if the file ends with a newline and inserts one if needed to prevent line concatenation issues.

SakuraIsayeki committed:
chore: Remove old whitelist check code

hakusaro commented:
@ACaiCat if you think the code is good, can you add a code review too?

hakusaro reviewed (changes requested):
I tested this and after adding 127.0.0.1 to the whitelist and using /reload, I was not able to whitelist myself.

Also, Terraria doesn't support IPv6.

hakusaro commented:
![CleanShot 2025-05-25 at 12 00 39](https://github.com/user-attachments/assets/3c46644e-8af5-496c-af77-195c834d4f64)

ACaiCat commented:
> @ACaiCat if you think the code is good, can you add a code review too?

Maybe you should find someone more professional, but it's not me.

hakusaro commented:
> Maybe you should find someone more professional, but it's not me.

You are professional

hakusaro commented:
Also, if the whitelist is blank, I think it would be fine to populate it with the example whitelist instead of leaving it blank. e.g., if whitelist is disabled, and is empty, add the default config to it.

hakusaro commented on TShockAPI/Whitelist.cs:
```diff
@@ -0,0 +1,270 @@
+				File.Delete(tempFile);
+			}
+
+            return removed;
```
![CleanShot 2025-05-25 at 12 04 54](https://github.com/user-attachments/assets/f9395a6c-d808-4c00-bdf9-f45e29d4af9b)


Indentation

SakuraIsayeki replied:
_May one day my Rider install be finally able to stop this lunatic disorder of switching between tabs and spaces on a whim..._

hakusaro replied:
https://www.sublimetext.com/

sgkoishi commented on TShockAPI/Commands.cs:
```diff
@@ -1742,12 +1742,8 @@ void BanDetails()
 		private static void Whitelist(CommandArgs args)
 		{
-			if (args.Parameters.Count == 1)
+			if (args.Parameters is [{ } ip] && TShock.Whitelist.AddToWhitelist(ip))
```
Perhaps show something even when not added (already existing)?

sgkoishi commented on TShockAPI/Whitelist.cs:
```diff
@@ -0,0 +1,270 @@
+			return true;
+		}
+
+		if (!IPAddress.TryParse(TShock.Utils.GetRealIP(host), out IPAddress? ip))
```
The `GetRealIP` is IPv4 only and will breaks IPv6... while Terraria doesn't support v6 🤣

sgkoishi commented on TShockAPI/Whitelist.cs:
```diff
@@ -0,0 +1,270 @@
+		{
+			// If we reach here, the line is not a valid IP address or network.
+			// We could throw this, but for now we just log and ignore it.
+			TShock.Log.Warn($"Invalid whitelist entry at line {line}: \"{content.ToString()}\", skipped");
```
Not a big problem but a line with just whitespace will ended up with a warning

ACaiCat commented:
> I tested this and after adding 127.0.0.1 to the whitelist and using /reload, I was not able to whitelist myself.
> 
> Also, Terraria doesn't support IPv6.

It seems can not be reloaded through `/reload` command (Need restart whole server)

sgkoishi commented on TShockAPI/Whitelist.cs:
```diff
@@ -0,0 +1,270 @@
+		=> _whitelistNetworks.Remove(network)
+		&& RemoveLine(network.ToString());
+
+	private bool AddLine(scoped ReadOnlySpan<char> content)
```
Both the `AddLine` and `RemoveLine` don't sync the in-memory list with the file, but instead add or remove from each separately. Not sure if this is intended, but sounds weird?

SakuraIsayeki replied:
Fully intended, Add/RemoveLine is only intended to combine the logic for file ops. They're internal calls destined to be used upstream by public-facing methods 😌

sgkoishi replied:
Yes, I'm referring to the public methods: for example, the whitelist in memory is empty and the file is `127.0.0.1` and `10.0.0.1`. 
Attempt to call the `public RemoveFromWhitelist("127.0.0.1")` will **not** remove `127.0.0.1` from the file (because it was not in memory), nor load `10.0.0.1` from the file. Similarly, `Add(a.b.c)` will not add it to the file if a.b.c is already in the memory.


(Due to the use of short-circuit logical AND `&&`)

SakuraIsayeki replied:
True point, I actually argued the use of the in-memory list being the source of truth for validation purposes, especially with dupes or any invalid notations in-file.
Stood to reason that, if duplicate (IIRC HashSet doesn't allow for duplicate values), adding to the list would fail, ergo shouldn't write to file. Similarly, pointless attempting to remove from file some value that isn't there or wasn't parsed.

Either I've gone crazy (that's what I get for missing my morning coffee...), or perhaps my logic wasn't very clear. Does this shed a bit more light as to why the && shorting ? Would you agree with this point, or would you like me to go with & operators instead to force the file ops in case of failures?

sgkoishi replied:
It works great when the memory and the file were syncing, but unintuitive when they are already **desynced** - might be kind of edge case:
* Run the server with default config
* Cleared the file content of whitelist.txt
* `/whitelist 127.0.0.1`
* `/reload` or restart the server
People would probably expect 127.0.0.1 to be whitelisted, but actually no it wasn't

SakuraIsayeki replied:
At this point, it does beg the question of file-watching for changes to sync the whitelist... 

I'll give it a shot.

sgkoishi replied:
Really edge case and I'm ok if we just keep the current behaviour - just a note in case someone encountered

hakusaro replied:
Just my 2c but this is why the original implementation, written by yours truly, is basic and just uses the file as a source of truth rather than syncing data structures.

SakuraIsayeki replied:
Aight, your call. I'm okay going the extra step and implementing a watcher, but would kinda defeat the purpose of a reload, I feel. 
Still, very thorough work, thanks for pointing out the edge case.

hakusaro replied:
I mean you don't have to implement a watcher -- you could just synchronize the data and write out the file if someone adds a new player to the list.

SakuraIsayeki replied:
Fair enough, let's just KISS it, I'll change the short-circuit operators.

SakuraIsayeki committed:
style: Fix indentations and newlines

SakuraIsayeki committed:
refactor: Improve whitelist command validation and feedback

Enhances the whitelist command to correctly handle parameter input, provides success or error messages based on the addition outcome.

SakuraIsayeki committed:
fix: Remove IPv6 whitelist support

Forgot to check if Terraria itself actually DOES support IPv6...

According to a few researches, latest hotfixes have grazed the subject. Hopefully we'll get IPv6 soon 🤞🏻

SakuraIsayeki committed:
feat: Add whitelist reload functionality

Introduces a method to reload whitelist data from file, allowing updates without restarting the server.

SakuraIsayeki committed:
Merge remote-tracking branch 'upstream/general-devel' into feature/whitelist

sgkoishi commented on TShockAPI/Whitelist.cs:
```diff
@@ -0,0 +1,291 @@
+		}
+
+		// HACK: Terraria doesn't support IPv6 yet, so we can't check for it.
+		// Remove once TShock supports IPv6.
```
This branch will never be true unless both underlying TShock and Terraria change - but let's just leave it without changing back and forth on these minor points...

hakusaro replied:
It *was* briefly supported in a released version of Terraria, but support was removed when an unexpected bug happened. I don't know if it will be re-added in a future update. It was live for about 8 hours before the hotfix removed it.

SakuraIsayeki replied:
Added this on the offchance Relogic does actually (re-)release IP6 support in a near future. Let's keep it there, if no one objects too sternly on it :)

sgkoishi reviewed (commented):
Other note:
* There's a submodule update that actually reverts 57039ed002, probably unwanted.
* The include of `.idea` folder?

SakuraIsayeki commented:
> Other note:
> 
> * There's a submodule update that actually reverts 57039ed002, probably unwanted.
> 
> * The include of `.idea` folder?

Oops, indeed unwanted changes. Lemme fix that right away 😅

hakusaro commented on .idea/.idea.TShock/.idea/.gitignore:
```diff
@@ -0,0 +1,13 @@
+/httpRequests/
+# Datasource local storage ignored files
+/dataSources/
+/dataSources.local.xml
```
```suggestion
.idea
```

SakuraIsayeki committed:
fix: Fix unwanted changes

Oops, should've paid more attention ^^'

SakuraIsayeki committed:
fix: use bitwise AND instead of logical AND for whitelist updates

Switches to bitwise AND operator to correctly combine add/remove operations with line updates.

ACaiCat commented on TShockAPI/Commands.cs:
```diff
@@ -1742,13 +1742,20 @@ void BanDetails()
+			}
+			else
+			{
+				args.Player.SendErrorMessage(GetString($"Invalid Whitelist syntax. Refer to {Specifier}whitelist help for details on how to use the {Specifier}whitelist command"));
```
It seems that the `/whitelist help` subcommand doesn't exist.

ACaiCat commented:
I test it on my win server, `/whitelist` command can't work.I am sure that no other process was using `whitelist.txt`
### Log:
```
2025-05-30 21:28:08 - Utils: INFO: Server execute: /whitelist 221.12.2.1。
2025-05-30 21:28:08 - Command: ERROR: System.IO.IOException: The process cannot access the file 'C:\Users\Administrator\Desktop\a\tshock\whitelist.txt' because it is being used by another process.
   at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options) in System.Private.CoreLib.dll:token 0x60001ce+0xa7
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode) in System.Private.CoreLib.dll:token 0x60001cd+0x6
   at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode) in System.Private.CoreLib.dll:token 0x6008cfe+0x14
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize, Nullable`1 unixCreateMode) in System.Private.CoreLib.dll:token 0x6008cd5+0xa
   at System.IO.FileInfo.OpenRead() in System.Private.CoreLib.dll:token 0x6008742+0x0
   at TShockAPI.Whitelist.AddLine(ReadOnlySpan`1 content) in /home/runner/work/TShock/TShock/TShockAPI/Whitelist.cs:line 227
   at TShockAPI.Whitelist.AddToWhitelist(IPAddress ip) in /home/runner/work/TShock/TShock/TShockAPI/Whitelist.cs:line 203
   at TShockAPI.Whitelist.AddToWhitelist(ReadOnlySpan`1 ip) in /home/runner/work/TShock/TShock/TShockAPI/Whitelist.cs:line 175
   at TShockAPI.Commands.Whitelist(CommandArgs args) in /home/runner/work/TShock/TShock/TShockAPI/Commands.cs:line 1747
   at TShockAPI.Command.Run(CommandArgs args) in /home/runner/work/TShock/TShock/TShockAPI/Commands.cs:line 158
```

SakuraIsayeki commented:
> I test it on my win server, `/whitelist` command can't work.I am sure that no other process was using `whitelist.txt`

By any chance, would you have the repro steps for this one ? Because I can't repro the issue.

ACaiCat commented:
> By any chance, would you have the repro steps for this one ? Because I can't repro the issue.

Just try to add a ip through `/whitelist 192.168.1.2`

ACaiCat commented:
@THEXN tested it too and got the same error.
![c34ee5cc3b502444c2953575df78f4d8](https://github.com/user-attachments/assets/f43aadf6-8cd6-4323-b0ff-c901e89a784c)

SakuraIsayeki committed:
fix: Correct fs handling in addline

SakuraIsayeki commented:
Okay. Seems I derped handling the file streams. 
Now fixed, give it a go on last commit :)

ACaiCat commented:
It works now.
![image](https://github.com/user-attachments/assets/957fa328-7683-4b7f-8e3a-8976e52bcf4c)

ACaiCat commented on TShockAPI/Whitelist.cs:
```diff
@@ -0,0 +1,291 @@
+		{
+			// If we reach here, the line is not a valid IP address or network.
+			// We could throw this, but for now we just log and ignore it.
+			TShock.Log.Warn($"Invalid whitelist entry at line {line}: \"{content.ToString()}\", skipped");
```
Adding GetString might be better
```suggestion
			TShock.Log.Warn(GetString($"Invalid whitelist entry at line {line}: \"{content.ToString()}\", skipped"));
```

SakuraIsayeki committed:
refactor: improve whitelist command handling

Add a warning message when IPv6 addresses are used in whitelist commands, as IPv6 support isn't implemented yet, and update error message for better clarity on correct command usage

SakuraIsayeki committed:
fix: localize logging messages for IPv6 support and invalid entries

Switches to localized strings for IPv6 warnings and invalid whitelist entry logs.

hakusaro committed:
Merge branch 'general-devel' into feature/whitelist
